### PR TITLE
DDT-1218 - update to 3.0 spec

### PR DESCRIPTION
**Pull request by FinnAtAsseco**

### CODE REVIEW

File: tnt-persistence/src/main/java/org/dcsa/tnt/persistence/repository/specification/EventCacheSpecification.java [hunk 1]
```diff
@@ -55,6 +68,9 @@ public static Specification<EventCache> withFilters(final EventCacheFilters filt
         predicates.add(root.get(EventCache_.EVENT_TYPE).in(filters.eventType));
       }
 
+      /*
+      TODO rewrite/update these filters for 3.0 spec filters
```
Comment (FinnAtAsseco) [inline tnt-persistence/src/main/java/org/dcsa/tnt/persistence/repository/specification/EventCacheSpecification.java:72]: Created https://dcsa.atlassian.net/browse/DDT-1265 to update filters and integration tests

File: tnt-integration-tests/src/main/java/org/dcsa/tnt/itests/v2/GetEventsIT.java [hunk 2]
```diff
@@ -98,15 +97,18 @@ public void getTransportEvents() {
         .jsonPath()
         .getList(".", EventTO.class);
 
-    events.forEach(e -> assertInstanceOf(TransportEventTO.class, e));
+    events.forEach(e -> assertInstanceOf(TransportEventPayloadTO.class, e.payload()));
   }
 
+  /*
+  TODO rewrite/update these tests for 3.0 spec filters
```
Comment (FinnAtAsseco) [inline tnt-integration-tests/src/main/java/org/dcsa/tnt/itests/v2/GetEventsIT.java:104]: Created https://dcsa.atlassian.net/browse/DDT-1265 to update filters and integration tests

File: tnt-application/src/main/java/org/dcsa/tnt/controller/EventController.java [hunk 3]
```diff
@@ -54,60 +71,112 @@ public List<EventTO> findEvents(
...
+
+    return Arrays.stream(sort.split(","))
+      .map(String::trim)
+      .map(sortField -> {
+        String[] fieldAndDirection = sortField.split(":");
+        String actualSortField = fieldAndDirection[0];
+        Direction direction = Direction.ASC;
+
+        if (!sortableFields.contains(actualSortField)) {
+          throw ConcreteRequestErrorMessageException.invalidInput("Cannot sort on '" + actualSortField + "'");
```
Comment (nt-as) [inline tnt-application/src/main/java/org/dcsa/tnt/controller/EventController.java:167]: I think `ConcreteRequestErrorMessageException.invalidQuery("sort", ...)` would be better (more specific)

File: tnt-application/src/main/java/org/dcsa/tnt/controller/EventController.java [hunk 4]
```diff
@@ -54,60 +71,112 @@ public List<EventTO> findEvents(
...
+        }
+
+        if (fieldAndDirection.length == 2) {
+          direction = Direction.fromOptionalString(fieldAndDirection[1])
+            .orElseThrow(() -> ConcreteRequestErrorMessageException.invalidInput(
+              "'" + fieldAndDirection[1] + "' is not a valid direction"));
+        } else if (fieldAndDirection.length > 2) {
+          throw ConcreteRequestErrorMessageException.invalidInput("'" + sortField + "' is not valid");
```
Comment (nt-as) [inline tnt-application/src/main/java/org/dcsa/tnt/controller/EventController.java:175]: I think `ConcreteRequestErrorMessageException.invalidQuery("sort", ...)` would be better (more specific)

File: tnt-application/src/main/java/org/dcsa/tnt/controller/EventController.java [hunk 5]
```diff
@@ -54,60 +71,112 @@ public List<EventTO> findEvents(
...
+        Direction direction = Direction.ASC;
+
+        if (!sortableFields.contains(actualSortField)) {
+          throw ConcreteRequestErrorMessageException.invalidInput("Cannot sort on '" + actualSortField + "'");
+        }
+
+        if (fieldAndDirection.length == 2) {
+          direction = Direction.fromOptionalString(fieldAndDirection[1])
+            .orElseThrow(() -> ConcreteRequestErrorMessageException.invalidInput(
+              "'" + fieldAndDirection[1] + "' is not a valid direction"));
```
Comment (nt-as) [inline tnt-application/src/main/java/org/dcsa/tnt/controller/EventController.java:173]: I think `ConcreteRequestErrorMessageException.invalidQuery("sort", ...)` would be better (more specific)

File: tnt-service/src/main/java/org/dcsa/tnt/service/mapping/transferobject/LocationMapper.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+      FacilityCodeListProvider facilityCodeListProvider;
+      if (facility.facilitySMDGCode() != null) {
+        facilityCode = facility.facilitySMDGCode();
+        facilityCodeListProvider = FacilityCodeListProvider.SMDG;
+      } else if (facility.facilityBICCode() != null) {
+        facilityCode = facility.facilityBICCode();
+        facilityCodeListProvider = FacilityCodeListProvider.BIC;
+      } else {
+        log.error("Facility {} has neither SMDG code nor BIC code", facility.id());
+        return null;
```
Comment (nt-as) [inline tnt-service/src/main/java/org/dcsa/tnt/service/mapping/transferobject/LocationMapper.java:42]: I think this warrants an exception (`IllegalArgumentException`)

File: tnt-service/src/main/java/org/dcsa/tnt/service/mapping/transferobject/LocationMapper.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        .facilityCodeListProvider(facilityCodeListProvider)
+        .build();
+    } else if (location.UNLocationCode() != null) {
+      return UNLocationLocationTO.builder()
+        .locationName(location.locationName())
+        .UNLocationCode(location.UNLocationCode())
+        .build();
+    } else {
+      log.error("Location {} has neither address, facility nor unLocation", location.id());
+      return null;
```
Comment (nt-as) [inline tnt-service/src/main/java/org/dcsa/tnt/service/mapping/transferobject/LocationMapper.java:57]: Ditto (`throw new IllegalArgumentException(...)` here please)

File: tnt-transfer-obj/src/main/java/org/dcsa/tnt/transferobjects/VesselTO.java [hunk 8]
```diff
@@ -7,17 +7,20 @@
...
+  String operatorCarrierCode, // <-
+  String operatorCarrierCodeListProvider // <-
+  /*
   CarrierTO vesselOperatorCarrier,
   Boolean isDummy,
   Float length,
   Float width,
   VesselType type,
   DimensionUnit dimensionUnit
+   */
```
Comment (nt-as) [inline tnt-transfer-obj/src/main/java/org/dcsa/tnt/transferobjects/VesselTO.java:23]: Left-over code or a TODO?

File: tnt-application/src/main/java/org/dcsa/tnt/controller/EventController.java (hunk 3 above)
Comment (FinnAtAsseco) [reply tnt-application/src/main/java/org/dcsa/tnt/controller/EventController.java]: Fixed

File: tnt-application/src/main/java/org/dcsa/tnt/controller/EventController.java (hunk 5 above)
Comment (FinnAtAsseco) [reply tnt-application/src/main/java/org/dcsa/tnt/controller/EventController.java]: Fixed

File: tnt-service/src/main/java/org/dcsa/tnt/service/mapping/transferobject/LocationMapper.java (hunk 7 above)
Comment (FinnAtAsseco) [reply tnt-service/src/main/java/org/dcsa/tnt/service/mapping/transferobject/LocationMapper.java]: Fixed

File: tnt-transfer-obj/src/main/java/org/dcsa/tnt/transferobjects/VesselTO.java (hunk 8 above)
Comment (FinnAtAsseco) [reply tnt-transfer-obj/src/main/java/org/dcsa/tnt/transferobjects/VesselTO.java]: Leftover, forgot to properly map vessels. Fixed now

File: tnt-service/src/main/java/org/dcsa/tnt/service/mapping/transferobject/LocationMapper.java (hunk 6 above)
Comment (FinnAtAsseco) [reply tnt-service/src/main/java/org/dcsa/tnt/service/mapping/transferobject/LocationMapper.java]: Fixed

**Review by nt-as — APPROVED** [review]

**Comment by nt-as** [comment]
@pimg / @douradofrazer : This is now ready for review. As I have written a considerable amount of changes on this PR, I feel one of you should also review the branch.

Note that this PR depends on https://github.com/dcsaorg/DCSA-Information-Model/pull/371

**Review by douradofrazer — APPROVED** [review]

**Pull request merged**